vault: add page

fluxw42 opened the pull request:
----
<!-- Thank you for sending a PR! -->
<!-- Please perform the following checks and check all the boxes that apply. -->
<!-- If a particular point is not applicable to your PR,
     strike-through the line by wrapping the text in ~~double tildes~~. -->
<!-- If your PR does not create or edit a command page (e.g. README edits, etc.)
     you can simply remove the entire checklist. -->

- [x] The page (if new), does not already exist in the repo.

- [x] The page (if new), has been added to the correct platform folder:  
      `common/` if it's common to all platforms, `linux/` if it's Linux-specific, and so on.

- [x] The page has 8 or fewer examples.

- [x] The PR is appropriately titled:  
      `<command name>: add page` for new pages, or `<command name>: <description of changes>` for pages being edited

- [x] The page follows the [contributing](https://github.com/tldr-pages/tldr/blob/master/CONTRIBUTING.md) guidelines 

This should resolve #1186

sbrl commented on pages/common/vault.md:
```diff
@@ -0,0 +1,27 @@
+
+`vault unseal {{key-share-x}}`
+
+- Authenticate client against vault, using an authentication token:
```
What precisely is meant by "authenticate client against vault"?

fluxw42 replied:
HashiCorps Vault is a server or daemon process. You have to authenticate the client process 'vault' before you can do any key-related actions.

sbrl replied:
Ahhh ok

sbrl commented on pages/common/vault.md:
```diff
@@ -0,0 +1,27 @@
+
+`vault read {{secret/hello}}`
+
+- Seal the vault:
```
Perhaps "seal the vault again"? Just to make it clear :P

fluxw42 replied:
Done.

sbrl replied:
Awesome :smiley_cat:

sbrl commented on pages/common/vault.md:
```diff
@@ -0,0 +1,27 @@
+
+`vault init -key-shares={{5}} -key-threshold={{2}}`
+
+- Unseal a vault, by providing one of the keyshares. Repeat with necessary key-shares until unsealed:
```
Don't think that the comment is needed here.

fluxw42 replied:
Ok, I'll make this shorter.

sbrl replied:
Oops  I meant comma, @Florent2! The comment is needed lol

fluxw42 replied:
Haha :-) Well, the *comment* is back but the *comma* is gone.

fluxw42 replied:
Haha :-) Well, the *comment* is back but the *comma* is gone.

agnivade commented on pages/common/vault.md:
```diff
@@ -0,0 +1,27 @@
+
+- Store a new secret in the vault:
+
+`vault write {{secret/hello}} value={{world}}`
```
Is the "/" necessary ? Can it be just "hello" ?

fluxw42 replied:
Vault uses different back-ends (secret, auth, audit, ..), which are modeled as a kind of file system. The 'secret/*' prefix in the example I provided points to their generic back-end to read and write arbitrary secrets.

You can mount other types of back-ends, eg. AWS, PKI, SSH, TOTP, ... but that would be too much to demonstrate in a tldr.

agnivade replied:
I see. So I recommend changing the description to `Store a new secret in the vault using the generic back-end called secret`

And move out the "secret" part out of the tokens- `vault write secret{{/hello}} value={{world}}`

fluxw42 replied:
Makes sense: Fixed!

agnivade commented on pages/common/vault.md:
```diff
@@ -0,0 +1,27 @@
+# vault
+
+> HashiCorp Vault, a tool for managing secrets.
```
Can we remove "HashiCorp Vault" ? It does not add any value to the explanation.

fluxw42 replied:
Would it be better if I changed it like this?

    > A CLI to interact with HashiCorp Vault

The 'vault' tool does nothing on it's own, you only use it to interact with a Vault server (locally or remotely).

agnivade commented:
Looks okay to me. However, I feel that the language can still be improved a bit. Waiting for @waldyrious' approval.

sbrl commented:
Oops! Looks like  mistyped in my comment above. See the inline comment.

waldyrious commented on pages/common/vault.md:
```diff
@@ -0,0 +1,27 @@
+
+> A CLI to interact with HashiCorp Vault.
+
+- Create a new vault, requiring at least 2 out of 5 keyshares to unseal:
```
I'm a little wary of starting out with this much jargon from the get-go ("keyshares"? "unseal"?).

Would it be possible to start with a simpler example (say, just `vault init`, relying on the defaults for the parameters)?

fluxw42 replied:
Using `vault init` would work, I'll simplify the example. This uses a default of 5 key shares requiring 3 of them to unseal.

waldyrious commented on pages/common/vault.md:
```diff
@@ -0,0 +1,27 @@
+
+`vault init -key-shares={{5}} -key-threshold={{2}}`
+
+- Unseal the vault by providing one of the keyshares. Repeat with necessary key-shares until unsealed:
```
What exactly does "unseal" mean in this context? Would it be possible to add a synonym or equivalent expression in parenthesis?

Also, we probably should use consistent spelling ("keyshares" or "key-shares").

fluxw42 replied:
When a vault is 'sealed' it means that the master key is missing to access the stored key- and config data. The master key is split up in a number of key shares (by default 5) and can only be reconstructed when a configured number of shares are combined again (by default 3 out of 5).

Unsealing is the process of combining these key shares to construct the original master key again. So when a vault is *sealed*, nobody can access it - not even authenticated clients -  since the master key is missing. When a vault is *unsealed*, it can be used by authenticated clients.

Would `unlocking` as a synonym for `unsealing` be better?

waldyrious commented on pages/common/vault.md:
```diff
@@ -0,0 +1,27 @@
+
+`vault unseal {{key-share-x}}`
+
+- Authenticate client against vault, using an authentication token:
```
I'm not really comfortable with the wording "Authenticate client against vault". Any change we could word it some other way?

waldyrious replied:
c.f. https://github.com/tldr-pages/tldr/pull/1475#discussion_r137553819

waldyrious commented on pages/common/vault.md:
```diff
@@ -0,0 +1,27 @@
+
+`vault auth {{authentication-token}}`
+
+- Store a new secret in the vault using the generic back-end called secret:
```
Please correct me if I'm reading this wrong, but is "secret" the name of the generic/default back-end? If so, maybe add quote marks around it to make that clearer, and perhaps a comma before "called".

waldyrious commented on pages/common/vault.md:
```diff
@@ -0,0 +1,27 @@
+
+`vault read secret/{{hello}}`
+
+- Seal the vault again:
```
Here we could take the opportunity that the description is short, to explain what seal/unseal does exactly. E.g., does this command unmount a folder, encrypt a file, mark stuff as read-only, or something else?

fluxw42 replied:
It deletes the master encryption key from memory, rendering all data inaccessible until someone *unseals* (or *unlocks*) the vault again.

waldyrious commented:
Overall looks good (nice work, @fluxw42, and welcome back -- hadn't seen you in a while!), but the descriptions can be improved IMO. I've left a few inline comments.

fluxw42 commented:
I think I've addressed all remarks that came up during the review. It's harder than expected to explain these basic commands in one concise sentence, there's a lot going on under the hood ...

Please take a look at it and let me know if I could still change something :)

waldyrious commented:
Looks good @fluxw42. We understand your pain, we have to deal with that struggle all the time when reviewing PRs. I think the page is quite acceptable as it is, but I'd only replace the single quote marks by double quotes, and use parenthesis for the synonyms, e.g.

> Seal (lock) the Vault server by removing the encryption key of the data store from memory:

Finally, you might want to throw in a few commas, to make the logical splits of clauses in the longer sentences more evident.

fluxw42 commented:
Should be fixed in 0ddfb4f: 
 * Single quotes `'` in description are now double quotes `"`.
 * Used parenthesis for *lock* and *unlock* synonyms.
 * Added commas where necessary, also removed two superfluous ones.

waldyrious commented:
@sbrl, the ball's on your court :smile:

sbrl commented:
Looks ok here :slightly_smiling_face:
